module/zfs/dmu_objset.c

```diff
@@ -1739,6 +1791,17 @@ dmu_objset_sync(objset_t *os, zio_t *pio, dmu_tx_t *tx)
 		    offsetof(dnode_t, dn_dirty_link[txgoff]));
 	}
 
+	/* sync_dnodes_finish_task executes zio and frees soa */
+	sync_objset_arg_t *soa = kmem_alloc(sizeof (*soa), KM_SLEEP);
+	soa->soa_os = os;
+	soa->soa_tx = tx;
+	soa->soa_zio = zio;
+	taskq_init_ent(&soa->soa_tq_ent);
+
+	/* sio is a child of the arc_write zio and parent of the sda_cio(s). */
+	zio_t *sio = zio_null(zio, os->os_spa, NULL,
```

Comment:
Is it okay that this zio completes before the zios that are created below in `sync_dnodes_task`? I thought that this needed to wait for those zios to complete.

Reply:
The child zios for this sio are all created before the code will call the zio_nowait(sio).  Those child zios are not _nowaited until after they have had their descendant zio hierarchies subequently created from sync_dnodes_task().

Reply:
fwiw I had this question too, as I've made this mistake before. But I think its ok, the wait ordering should be correct; the parent will wait in `zio_ready()` if it beats the children there.

Reply:
Code updated per discussions.